Added doc for etcd member garbage collection (#420)
gardener/etcd-backup-restore: doc/concepts/etcd_member_garbage_collection.md

## Superfluous member deletion in the ETCD cluster

With support for multi-node operation, there is an expectation that the cluster size can be dynamic. In such scenarios, a reduction in the cluster size should not result in any superfluous members or entries being left over.

> Such superfluour members/entries can have a adverse affect on the quorum size of the cluster and could result in a degradation of service.

Reviewer: ```suggestion
Such superfluour members/entries can have an adverse affect on the quorum size of the cluster and could result in a degradation of service.
```

As such, care has to be taken to periodically check and remove any superfluous entries.

There is the possibility of superfluous entries in the `members` array of the ETCD CRD as well as in the ETCD cluster for which there is no corresponding pod in the `StatefulSet` anymore.

> Keeping the `members` array on the ETCD CRD in check is the responsibility of `etcd-druid` whereas keeping the ETCD cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar

Reviewer: ```suggestion
Keeping the `members` array on the ETCD CRD in check is the responsibility of `etcd-druid` whereas keeping the ETCD cluster in check is the responsibility of the leading `etcd-backup-restore` sidecar.
```

> The scope of this document is the responsibility of the leading `etcd-backup-restore` sidecar to keep the ETCD cluster in check

Reviewer: ```suggestion
The purpose of this document is to describe the responsibility of the leading `etcd-backup-restore` sidecar to keep the ETCD cluster in check.
```

#### Example

> Below is as example of how we could end up with superfluous members in the ETCD cluster.

Reviewer: ```suggestion
Below is an example of how we could end up with superfluous members in the ETCD cluster.
```

This is a common scenario that could occur frequently and demonstrates the need for a garbage collectoin mechanism that detects and clears such superfluous entries automatically
```

> Suppose an ETCD cluster is scaled up from `3` to `5` and the newly added members were failing constantly due to insufficient resources and then if the ETCD cluster is scaled back down to `3`. The failing member pods may not have the chance to clean up their `member` entries (from the `members` array in the ETCD CRD as well as from the ETCD cluster) leading to superfluous members in the ETCD cluster that may have adverse effect on quorum of the cluster.

Reviewer: I just wondered what we should do if a etcd-member pod receives a SIGTERM which can either mean:
- The pod is removed/evicted
- A rolling update is happening
- The statefulset is scaled down

What we already consider today IIUC:
- Regular garbage collection of superfluous members.
- When a member pod starts, it checks if there is an "outdated" member entry which it is going to replace, i.e. it removes the entry and adds itself again.

:question: Do we also want to have a cleanup logic if SIGTERM arrives or do we solely want to rely on the mechanisms listed above?

Let's discuss this first and then we can adjust this example.

Reviewer: As discussed offline, we're going to go with a cleanup logic when SIGTERM arrives.

Can you please delete "from the `members` array in the ETCD CRD" as this is not the case anymore?

```

### Superfluous member 

> A member in the etcd cluster can be considered as superfluous only if the following three conditiond hold true

Reviewer: ```suggestion
A member in the etcd cluster can be considered as superfluous only if the following three conditions hold true:
```

1. It does not have a corresponding pod

> 2. It does not have an enter in the ETCD CRD member list (`status.members`)

Reviewer: I think this is not relevant anymore, right?

3. It does not have a corresponding etcd member lease

> It may not be necessary to verify all 3 conditions before calling a member superfluous. For instance, `status.members` is kept in sync with member leases by `etcd-druid`.
> Hence, checking the for the correspondng pod and the member entry in `status.members` is a sufficient check

Reviewer: ```suggestion
```
This is also not relevant anymore, right?

### Approach
- Superfluous members garbage collection process will only be triggered by the leading `etcd-backup-restore` sidecar

> - The garbage collection mechanish runs periodically

Reviewer: Can you describe when and how to influence this?
